### DISCUSSION TITLE
PostgreSQL foreign key constraint

### DISCUSSION THREAD

**Discussion by mshiner** [Budibase/budibase#5913] [Help] [answered]
I have two tables in my postgres DB as follows
```
CREATE TABLE IF NOT EXISTS public.gl_analysis
(
    id bigint NOT NULL DEFAULT nextval('gl_analysis_id_seq'::regclass),
    analysis character varying COLLATE pg_catalog."default" NOT NULL,
    CONSTRAINT glanalysis_pkey PRIMARY KEY (id)
)
```
and
```
CREATE TABLE IF NOT EXISTS public.gl_accounts
(
    id bigint NOT NULL DEFAULT nextval('gl_accounts_id_seq'::regclass),
    account character varying COLLATE pg_catalog."default" NOT NULL,
    actype character varying COLLATE pg_catalog."default" NOT NULL,
    description character varying COLLATE pg_catalog."default" NOT NULL,
    control boolean NOT NULL,
    header boolean NOT NULL,
    glanalysis_id smallint,
    CONSTRAINT glmaster_pkey PRIMARY KEY (id),
    CONSTRAINT glmaster_glanalysis_id_fkey FOREIGN KEY (glanalysis_id)
        REFERENCES public.gl_analysis (id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION,
    CONSTRAINT glmaster_actype_check CHECK (actype::bpchar = 'P'::bpchar OR actype::bpchar = 'B'::bpchar)
)
```

When I define the relationship between the tables in Budibase as follows

![Screenshot from 2022-05-16 17-43-47](https://user-images.githubusercontent.com/16088458/168642662-65d41e0b-14bd-4e66-8fe1-a3605243512f.png)

In the table view I do not see the analysis_id data in the 'created' metadata column gl_analysis in the gl_accounts table even though the data is there in the actual table column(glanalysis_id) For reference I can do the floowing select and all is well:

```
SELECT m.id,
    m.account,
    m.actype,
    m.description,
    m.control,
    m.header,
    m.glanalysis_id,
    a.analysis::text AS analysis
   FROM gl_accounts m
     LEFT JOIN gl_analysis a ON m.glanalysis_id = a.id;
```

Am I doing something wrong?

Thanks
M

ps... it'd be great if Budibase could introspect the db and surface the relationships from FK constraints as there are a lot of tables with a lot of relationships to define (perhaps something like Prisma does?)

**Comment by melohagan** [comment, marked as answer]
Hi @mshiner 

I was able to reproduce this issue - it's because the types of the ids don't match:

- gl_analysis (id bigint)
- gl_accounts (glanalysis_id smallint)

I changed *glanalysis_id* to a **bigint** and then the relationship worked.

**Reply by mshiner** [reply]
Interesting.... thanks for the heads up.

postgreSQL itself just does an internal (implicit) cast when doing the FK constraint as I believe it is set in the pg_catalog as a type equivalent.

The database in question, from the [uzERP open source ERP system](https://github.com/uzerpllp/uzerp) has a lot of this I think so I'll have to be careful.

**Reply by melohagan** [reply]
It could be good to add a feature request for implicit casts on FKs. Do you have a link to the pg_catalog were I could see a list of the equivalent types?

**Reply by mshiner** [reply]
Hi

Here ya go https://www.postgresql.org/docs/current/catalogs.html and https://www.postgresql.org/docs/current/catalog-pg-cast.html

I think some are 'internal' and not in pg_cast so maybe query that with the postgeSQL devs?

Regards
M

**Reply by mshiner** [reply]
Can I also ask..... why no database views?

The ERP system in question (uzERP) makes extensive use of views in the public schema to define the data 'collection' for use in the MVC framework it is built on. If I could see and use the views in budibase I wouldn't have to redefine them. 

Regards
M

**Reply by melohagan** [reply]
There are some basic views for the internal tables, but for the external datasources, the idea was that you could use custom queries to effectively create views: https://docs.budibase.com/docs/custom-queries

**Reply by mshiner** [reply]
Yep - the custom query thing is a good feature - for me (in my use case) it'd save me a bit of time if I could see and use the standard views in my db as it'd save a bit of time.... there are rather a lot of them :wink:

**Reply by mshiner** [reply]
Just to say though.... Budibase does look like the tool I've been looking for to add some value to the base ERP system.

M
